### DISCUSSION TITLE
Capturing whitespace and comments

### DISCUSSION THREAD

**Discussion by carueda** [zesterer/chumsky#542] [Q&A] [answered]
I'd like to ask about any "best practices" to capture whitespace/comments using Chumsky. What I've done in some project of mine (in Scala) is to capture any such elements at the lexical, token level, that is, each token instance captures not only its kind, lexeme and span per se, but also any immediately preceding whitespace/comments (this is a bit of a simplification, but hopefully enough for the matter at hand.)

Does that seem like a good strategy using Chumsky, or are there better approaches? Thanks.

**Comment by zesterer** [comment, marked as answer]
I would say this is a question that sits a little beyond chumsky (although chumsky should be more than capable of handling whatever solution you choose to go for). I assume that the goal here is to produce an AST from which the original source text can be reconstructed?

A few solutions I'd think would work nicely (note that I'm using the [`map_with` API](https://github.com/zesterer/chumsky/pull/537) that was added in the latest alpha release):

1) Store the whitespace that appears to the left and the right within the token itself:

```rs
enum TokenKind<a> {
    Fn,
    Ident(&'a str),
    ...
}

struct Token<'a> {
    kind: TokenKind<'a>,
    span: SimpleSpan,
    left_ws: &'a str,
    right_ws: &'a str,
}

// In the lexer (which is just a kind of parser, of course)

let token_kind = /* parser for `TokenKind`s */

let token = group((
    text::whitespace(),
    token_kind,
    text::whitespace(),
))
    .map_with(|(left_ws, kind, right_ws), e| Token { kind, left_ws, right_ws, span: e.span() });
```

2) Store two spans: one for the token, another for the token + the surrounding whitespace (meaning that recovering the whitespace for the token, given the input to index into, becomes easy):

```rs
enum TokenKind<a> {
    Fn,
    Ident(&'a str),
    ...
}

struct Token<'a> {
    kind: TokenKind<'a>,
    span: SimpleSpan,
    ws_span: SimpleSpan,
}

let token_kind = /* parser for `TokenKind`s */

let token = token_kind
    .map_with(|kind, e| (kind, e.span()))
    .padded()
    .map_with(|(kind, span), e| Token {
        kind,
        span,
        ws_span: e.span(),
    });
```

3) Store the string slice corresponding to both token and surrounding whitespace:

```rs
enum TokenKind<a> {
    Fn,
    Ident(&'a str),
    ...
}

struct Token<'a> {
    kind: TokenKind<'a>,
    token_and_ws: &'a str,
}

let token_kind = /* parser for `TokenKind`s */

let token = token_kind
    .padded()
    .map_with(|kind, e| Token {
        kind,
        token_and_ws: e.slice(),
    });
```

Which of these you use (or whether you use something else different entirely) is up to you and depends on the constraints/use-case you have. If nothing else, I hope this at least gives a sense of how one might approach this problem with chumsky.

**Reply by carueda** [reply]
That's great! Yes, use cases include not only reconstructing the original source, but also source re-formatting, and translation to other language where comments are to be retained. Yes, those definitely give me a very good sense of possible approaches. Thanks much, I really appreciate the detailed replies.
